core/io/image.cpp

```diff
@@ -2871,48 +2871,48 @@ Color Image::_get_color_at_ofs(const uint8_t *ptr, uint32_t ofs) const {
 void Image::_set_color_at_ofs(uint8_t *ptr, uint32_t ofs, const Color &p_color) {
 	switch (format) {
 		case FORMAT_L8: {
-			ptr[ofs] = uint8_t(CLAMP(p_color.get_v() * 255.0, 0, 255));
+			ptr[ofs] = (uint8_t)Math::round(p_color.get_v() * 255.0);
 		} break;
 		case FORMAT_LA8: {
-			ptr[ofs * 2 + 0] = uint8_t(CLAMP(p_color.get_v() * 255.0, 0, 255));
-			ptr[ofs * 2 + 1] = uint8_t(CLAMP(p_color.a * 255.0, 0, 255));
+			ptr[ofs * 2 + 0] = (uint8_t)Math::round(p_color.get_v() * 255.0);
+			ptr[ofs * 2 + 1] = (uint8_t)Math::round(p_color.a * 255.0);
 		} break;
 		case FORMAT_R8: {
-			ptr[ofs] = uint8_t(CLAMP(p_color.r * 255.0, 0, 255));
+			ptr[ofs] = (uint8_t)Math::round(p_color.r * 255.0);
 		} break;
 		case FORMAT_RG8: {
-			ptr[ofs * 2 + 0] = uint8_t(CLAMP(p_color.r * 255.0, 0, 255));
-			ptr[ofs * 2 + 1] = uint8_t(CLAMP(p_color.g * 255.0, 0, 255));
+			ptr[ofs * 2 + 0] = (uint8_t)Math::round(p_color.r * 255.0);
+			ptr[ofs * 2 + 1] = (uint8_t)Math::round(p_color.g * 255.0);
 		} break;
 		case FORMAT_RGB8: {
-			ptr[ofs * 3 + 0] = uint8_t(CLAMP(p_color.r * 255.0, 0, 255));
-			ptr[ofs * 3 + 1] = uint8_t(CLAMP(p_color.g * 255.0, 0, 255));
-			ptr[ofs * 3 + 2] = uint8_t(CLAMP(p_color.b * 255.0, 0, 255));
+			ptr[ofs * 3 + 0] = (uint8_t)Math::round(p_color.r * 255.0);
+			ptr[ofs * 3 + 1] = (uint8_t)Math::round(p_color.g * 255.0);
+			ptr[ofs * 3 + 2] = (uint8_t)Math::round(p_color.b * 255.0);
 		} break;
 		case FORMAT_RGBA8: {
-			ptr[ofs * 4 + 0] = uint8_t(CLAMP(p_color.r * 255.0, 0, 255));
-			ptr[ofs * 4 + 1] = uint8_t(CLAMP(p_color.g * 255.0, 0, 255));
-			ptr[ofs * 4 + 2] = uint8_t(CLAMP(p_color.b * 255.0, 0, 255));
-			ptr[ofs * 4 + 3] = uint8_t(CLAMP(p_color.a * 255.0, 0, 255));
+			ptr[ofs * 4 + 0] = (uint8_t)Math::round(p_color.r * 255.0);
+			ptr[ofs * 4 + 1] = (uint8_t)Math::round(p_color.g * 255.0);
+			ptr[ofs * 4 + 2] = (uint8_t)Math::round(p_color.b * 255.0);
+			ptr[ofs * 4 + 3] = (uint8_t)Math::round(p_color.a * 255.0);
 
 		} break;
 		case FORMAT_RGBA4444: {
 			uint16_t rgba = 0;
 
-			rgba = uint16_t(CLAMP(p_color.r * 15.0, 0, 15)) << 12;
-			rgba |= uint16_t(CLAMP(p_color.g * 15.0, 0, 15)) << 8;
-			rgba |= uint16_t(CLAMP(p_color.b * 15.0, 0, 15)) << 4;
-			rgba |= uint16_t(CLAMP(p_color.a * 15.0, 0, 15));
+			rgba = (uint16_t)Math::round(p_color.r * 15.0) << 12;
+			rgba |= (uint16_t)Math::round(p_color.g * 15.0) << 8;
+			rgba |= (uint16_t)Math::round(p_color.b * 15.0) << 4;
+			rgba |= (uint16_t)Math::round(p_color.a * 15.0);
 
 			((uint16_t *)ptr)[ofs] = rgba;
 
 		} break;
 		case FORMAT_RGB565: {
 			uint16_t rgba = 0;
 
-			rgba = uint16_t(CLAMP(p_color.r * 31.0, 0, 31));
-			rgba |= uint16_t(CLAMP(p_color.g * 63.0, 0, 33)) << 5;
-			rgba |= uint16_t(CLAMP(p_color.b * 31.0, 0, 31)) << 11;
+			rgba = (uint16_t)Math::round(p_color.r * 31.0);
+			rgba |= (uint16_t)Math::round(p_color.g * 63.0) << 5;
```

Comment:
Notice this while pressing "approve"... Typo, `63.0` where it should be `33.0`.

Reply:
Ah well the `63.0` vs `33` is from the original code. Might have been a bug, needs double checking.

I'm not sure about the clamp removal either, this does more than just rounding instead of flooring.

Reply:
I think you have to round then clamp.

Reply:
Pushed the changes myself.